### PR TITLE
Fixed broken links in README.md and some tests

### PR DESCRIPTION
**Pull request by jakbin**

### CODE REVIEW

**Comment by jakbin** [comment]
@fredysomy

**Comment by jakbin** [comment]
@Adwaith-Rajesh please review these changes .

**Review by Adwaith-Rajesh — COMMENTED** [review]
Did you run these test ???

**Comment by jakbin** [comment]
Yes I run . There is two error.

**Comment by Adwaith-Rajesh** [comment]
Pls Fix it, 👍

**Comment by Adwaith-Rajesh** [comment]
all the test should pass

**Comment by jakbin** [comment]
@Adwaith-Rajesh I done.

**Comment by Adwaith-Rajesh** [comment]
pls wait

**Comment by jakbin** [comment]
> pls wait

For what ?

**Comment by jakbin** [comment]
@Adwaith-Rajesh Now its cover all core function of db.

**Review by Adwaith-Rajesh — CHANGES_REQUESTED** [review]
```
>>> 1 == True
True
>>> 1 is True
False
```

File: tests/tests_json_database.py [hunk 1]
```diff
@@ -33,104 +32,128 @@ def test_database_add_many(tmpdir):
...
+        {"name": "test", "getbyfield": "row1"},
+        {"name": "pysondb", "getbyfield": "row2"},
+        {"name": "def23@c-py", "getbyfield": "row3"},
+        {"name": "stuff(py", "type": "GUI"}
+    ]
+    a.addMany(fixture)
+    assert (a.reSearch("name", "test"))[0]["name"] == fixture[0]["name"]
+    assert (a.reSearch("getbyfield", "row2"))[0]["getbyfield"] == fixture[1]["getbyfield"]
+    assert (a.reSearch("name", r"\w{3}\d{2}@c-py"))[0]["name"] == fixture[2]["name"]
+    # assert (a.reSearch("name", "stuff\(py"))[0]["name"] == fixture[3]["name"]          # this is not working
```
[the remark this answers is not in the document]
Comment (jakbin) [reply tests/tests_json_database.py:159]: its showing error - IndexError: list index out of range

File: tests/tests_json_database.py [hunk 2]
```diff
@@ -33,104 +32,128 @@ def test_database_add_many(tmpdir):
...
+
+def test_database_research(tmpdir):
+    file = tmpdir.join("test.db.json")
+    file.write(EMPTY_FIXTURE_STR)
+    a = getDb(file.strpath)
+    fixture = [
+        {"name": "test", "getbyfield": "row1"},
+        {"name": "pysondb", "getbyfield": "row2"},
+        {"name": "def23@c-py", "getbyfield": "row3"},
+        {"name": "stuff(py", "type": "GUI"}
```
Comment (Adwaith-Rajesh) [inline tests/tests_json_database.py:153]: `{"name": "test", "getbyfield": "row1"},` this line sets the schema of the DB to have two keys, that are, name and getByField. so when you add the last one `{"name": "stuff(py", "type": "GUI"}` this no longer compiles to the schema and thus is not added to the DB. and thats why you are getting an IndexError

**Review by Adwaith-Rajesh — CHANGES_REQUESTED** [review]

File: tests/tests_json_database.py (hunk 2 above)
Comment (jakbin) [reply tests/tests_json_database.py:153]: Thanks brother

**Comment by jakbin** [comment]
@Adwaith-Rajesh Please review [these](https://github.com/pysonDB/pysonDB/pull/69/commits/cd3ad5aa845870d6c115538bbd00f7a49471a57c) changes.

**Comment by Adwaith-Rajesh** [comment]
You also have to test whether the "CLI" works, like this.

**Comment by Adwaith-Rajesh** [comment]
you have to test the "CLI" like this
[](https://i.imgur.com/iFy8SLn.png)

**Comment by Adwaith-Rajesh** [comment]
```python
# cli.py
import argparse
import sys
from typing import Sequence


def main(argv: Sequence | None = None) -> int:

    parser = argparse.ArgumentParser()
    parser.add_argument("name")

    args = parser.parse_args(argv)
    if args.name == '':
        print("The name must not be empty", file=sys.stderr)
        return 1

    print(f"Hello {args.name}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

```

```python
# test_cli.py
import pytest
import cli


def test_main(capsys):
    cli.main(["name"])

    out, err = capsys.readouterr()
    assert out == "Hello name\n"
    assert err == ""


def test_main_empty_string(capsys):
    cli.main([""])

    out, err = capsys.readouterr()
    assert out == ""
    assert err == "The name must not be empty\n"

```

**Comment by Adwaith-Rajesh** [comment]
this is why the main cli function accepts and sequence of strings as an arg

**Comment by Adwaith-Rajesh** [comment]
And you are also missing a lot (A LOT) of edge cases in all the test that u've written.

**Comment by jakbin** [comment]
Thanks for your review. I will improve my code .

**Comment by jakbin** [comment]
@Adwaith-Rajesh But how to pass argument for this ?
```python
 
 ​def​ ​set_parser​(​argv​: ​Optional​[​Sequence​[​str​]] ​=​ ​None​) ​->​ ​int​: 
 ​    ​parser​ ​=​ ​argparse​.​ArgumentParser​(​description​=​"Command line utility for pysondb"​, ​epilog​=​example_uses​, ​formatter_class​=​argparse​.​RawDescriptionHelpFormatter​) 
 ​    ​subparsers​ ​=​ ​parser​.​add_subparsers​(​dest​=​"command"​) 
  
 ​    ​create_parser​ ​=​ ​subparsers​.​add_parser​(​"create"​, ​help​=​"Create a new database file"​) 
 ​    ​create_parser​.​add_argument​(​"file_name"​, ​help​=​"Name of the database file"​)
```

**Comment by Adwaith-Rajesh** [comment]
@jakbin sry for the delay

**Comment by Adwaith-Rajesh** [comment]
your test file will look smthg like this.

```python
import pytest

from pysondb.cli import set_parser


def test_main(capsys):
    set_parser(["arg1", "arg2"])

    out, err = capsys.readouterr()
    assert out == "Expected output"
    assert err == ""


def test_main_empty_string(capsys):
    set_parser(["some", "error"])

    out, err = capsys.readouterr()
    assert out == ""
    assert err == "Expected error"
```

**Comment by Adwaith-Rajesh** [comment]
I really don't why he created an extra function to write all of this, but I allowed it anyway.

**Comment by jakbin** [comment]
@Adwaith-Rajesh  How its is [this](https://github.com/pysonDB/pysonDB/pull/69/commits/547d3e2fe5f4ccc77e7c966f000839d44180fa50) ?

**Review by Adwaith-Rajesh — APPROVED** [review]
I will merge this to testing branch, we will check whether everything works and we will slowly update and merge it to master.

Thank you so so so much for you contribution, appreciate it.
